iceoryx_posh/source/roudi/service_registry.cpp

```diff
@@ -17,45 +17,149 @@
 
 #include "iceoryx_posh/internal/roudi/service_registry.hpp"
 
+#include <algorithm>
+
 namespace iox
 {
 namespace roudi
 {
-void ServiceRegistry::add(const capro::IdString_t& service, const capro::IdString_t& instance)
+cxx::expected<ServiceRegistry::Error> ServiceRegistry::add(const capro::ServiceDescription& serviceDescription) noexcept
 {
-    cxx::set::add(m_serviceMap[service].instanceSet, instance);
+    if (!serviceDescription.isValid())
+    {
+        return cxx::error<Error>(Error::SERVICE_DESCRIPTION_INVALID);
+    }
+
+    // Forbid duplicate service descriptions entries
+    for (auto& element : m_serviceDescriptionVector)
```

Comment:
We can avoid this linear search later, could be worth adding a todo. We already have search structures in place that allow us to find quickly whether the service already exists, that is the major point of this data structure itself. If so, we return but not with an error (see below).